text/2-aurgy.md

```diff
@@ -0,0 +1,135 @@
+
+#### Creating and managing a group/lobby
+
+In this project, users should be able to join a private, protected group. We could probably
+accomplish this using a standard lobby system: one person acts as the host and creates the
+lobby and receives a unique code; users who then enter this code can enter that particular
+lobby.
+
+We could probably set something like this up with a simple document database. Whenever a host
+creates a lobby, a new document is created with a unique identifying code. Then, if another
+user enters that code, their ID is added to the document. The document might look something
+like this:
+```json
+"<unique id>": {
+  "users": [
+    "<id>", "<id>", ...
+  ]
+}
+```
+
+In the interim, something like Firebase Firestore should be enough. It might be a good idea
+to switch to something else (like MongoDB perhaps) if we want to worry about scaling and
+whatnot.
+
+We could achieve persistent groups by just keeping these unique identifiers persistent, i.e.
+not deleting them. This way, the host and any other users can re-enter groups they are a part
+of by just entering their unique code.
+
+#### Generating a playlist
+
+Here's the meat of the project. We would make heavy use of the [Spotify API] and all of the
+information available about individual tracks here.
+
+First, we need a list of themes that the user can choose from. Some suggestions are "road trip",
+"party", or "kickback". Then, we would have to correlate each theme with a set of track features,
+such as genre.
+
+We would also need to understand the users' music preferences. Again, we would pull things like
+favorite (recent?) genres and figure out areas of strongest intersection between users.
+
+Finally, we want to take the features prescribed by the desired theme and the features that are
+most common among users, and pull songs that match as many of these features as possible.
+
+Once the playlist is generated, we can add it to the host's library.
+
+Playlist generation would of course be performed on the server side.
+
+[Spotify API]: https://developer.spotify.com/documentation/web-api/
+
+#### Playing music from the playlist
+
+Spotify offers a [playback SDK] to play music from Spotify on the client side. Combined with the
+[web API] for playback controls, we have a way for music to be played directly from our application.
+
+[playback SDK]: https://developer.spotify.com/documentation/web-playback-sdk/
+[web API]: https://developer.spotify.com/documentation/web-api/
+
+### What are the requirements for this project?
+
+The PWA itself will probably be built in React (possibly Next.js? I don't think it's necessary
+for this use case but @BryanPan342 what do you think?), and the backend will be Node.js-based.
+If we proceed with using Firestore for handling user lobbies to start with, then anyone with
+Firebase experience would be handy.
+
+We will probably need OCI to host a server for generating playlists, since it makes more sense
```

Comment:
Alternate option is to use Firebase Functions for serverless - they have a very nice way of converting an express server into smth that plays nice with serverless with very little work. Additionally doing so would probably allow us to very easily play nice with any auth middleware we may or may not need.

Reply:
That's a good point and I remember we discussed serverless as well and I'm not sure why i forgot about it completely but I'll throw in serverless